testutil/ciao-down/cloudinit.go

```diff
@@ -0,0 +1,427 @@
+ - curl -X PUT -d "Setting git user.name" 10.0.2.2:{{.HTTPServerPort}}
+ - sudo -u {{.User}} git config --global user.name "{{.GitUserName}}"
+{{end}}
+{{if len .GitEmail}}
+ - curl -X PUT -d "Setting git user.email" 10.0.2.2:{{.HTTPServerPort}}
+ - sudo -u {{.User}} git config --global user.email {{.GitEmail}}
+{{end}}
+ - curl -X PUT -d "FINISHED" 10.0.2.2:{{.HTTPServerPort}}
+
+users:
+  - name: {{.User}}
+    gecos: CIAO Demo User
+    lock-passwd: true
+    shell: /bin/bash
+    passwd: $1$vzmNmLLD$04bivxcjdXRzZLUd.enRl1
+    sudo: ALL=(ALL) NOPASSWD:ALL
+    ssh-authorized-keys:
+    - {{.PublicKey}}
+`
+
+const metaDataTemplate = `
+{
+  "uuid": "ddb4a2de-e5a5-4107-b302-e845cecd7613",
+  "hostname": "singlevm"
+}
+`
+
+// TODO: Code copied from launcher.  Needs to be moved to qemu
+
+func createCloudInitISO(ctx context.Context, instanceDir string, userData, metaData []byte) error {
+	configDrivePath := path.Join(instanceDir, "clr-cloud-init")
+	dataDirPath := path.Join(configDrivePath, "openstack", "latest")
+	metaDataPath := path.Join(dataDirPath, "meta_data.json")
+	userDataPath := path.Join(dataDirPath, "user_data")
+	isoPath := path.Join(instanceDir, "config.iso")
+
+	defer func() {
+		_ = os.RemoveAll(configDrivePath)
+	}()
+
+	err := os.MkdirAll(dataDirPath, 0755)
+	if err != nil {
+		glog.Errorf("Unable to create config drive directory %s", dataDirPath)
+		return err
+	}
+
+	err = ioutil.WriteFile(metaDataPath, metaData, 0644)
+	if err != nil {
+		glog.Errorf("Unable to create %s", metaDataPath)
+		return err
+	}
+
+	err = ioutil.WriteFile(userDataPath, userData, 0644)
+	if err != nil {
+		glog.Errorf("Unable to create %s", userDataPath)
+		return err
+	}
+
+	cmd := exec.CommandContext(ctx, "xorriso", "-as", "mkisofs", "-R", "-V", "config-2",
+		"-o", isoPath, configDrivePath)
+	err = cmd.Run()
+	if err != nil {
+		return fmt.Errorf("Unable to create cloudinit iso image %v", err)
+	}
+
+	return nil
+}
+
+func buildISOImage(ctx context.Context, instanceDir string, ws *workspace) error {
+	udt := template.Must(template.New("user-data").Parse(userDataTemplate))
+	var udBuf bytes.Buffer
+	err := udt.Execute(&udBuf, ws)
+	if err != nil {
+		return fmt.Errorf("Unable to execute user data template : %v", err)
+	}
+
+	mdt := template.Must(template.New("meta-data").Parse(metaDataTemplate))
+
+	var mdBuf bytes.Buffer
+	err = mdt.Execute(&mdBuf, ws)
+	if err != nil {
+		return fmt.Errorf("Unable to execute user data template : %v", err)
+	}
+
+	return createCloudInitISO(ctx, instanceDir, udBuf.Bytes(), mdBuf.Bytes())
+}
+
+// TODO: Code copied from launcher.  Needs to be moved to qemu
+
+func createRootfs(ctx context.Context, backingImage, instanceDir string) error {
+	vmImage := path.Join(instanceDir, "image.qcow2")
+	if _, err := os.Stat(vmImage); err == nil {
+		_ = os.Remove(vmImage)
+	}
+	params := make([]string, 0, 32)
+	params = append(params, "create", "-f", "qcow2", "-o", "backing_file="+backingImage,
+		vmImage, "60000M")
```

Comment:
This 60G number should be a variable and documented.  60G is a significant chunk.  It's big enough to make a VM able to run a long time with a bunch of ceph or docker left over trash, and then quietly die a few months down the line inexplicably.  But it also may be more than somebody wants to give to their Ciao development slowly incrementally filling up their disk.  I think single VM needs close to 20G for a full-ish Ubuntu install and various downloaded images and their copy into ceph, but am not sure the specific number.  Documenting this well will help the users and a config knob or greppable variable name will help them deal with future issues.